src/cider/exec/operator/join/CiderLinearProbingHashTable.h

```diff
@@ -228,57 +240,27 @@ class LinearProbeHashTable {
 
  private:
   template <typename K, typename... Args>
-  bool emplace_impl(const K& key, Args&&... args) {
-    reserve(size_ + 1);
-    for (size_t idx = key_to_idx(key);; idx = probe_next(idx)) {
-      if (!buckets_[idx].first.is_not_null) {
-        buckets_[idx].second = mapped_type(std::forward<Args>(args)...);
-        buckets_[idx].first.key = key;
-        buckets_[idx].first.is_not_null = true;
-        size_++;
-        return true;
-      } else if (key_equal()(buckets_[idx].first.key, key)) {
-        buckets_[idx].first.duplicate_num++;
-      }
-    }
-    return false;
-  }
+  bool emplace_impl(const K& key, Args&&... args);
 
   template <typename K>
-  folly::fbvector<mapped_type> find_impl(const K& key) {
-    folly::fbvector<mapped_type> vec;
-    int duplicate_num = 0;
-    for (size_t idx = key_to_idx(key);; idx = probe_next(idx)) {
-      if (key_equal()(buckets_[idx].first.key, key) && buckets_[idx].first.is_not_null) {
-        if (vec.size() == 0) {
-          duplicate_num = buckets_[idx].first.duplicate_num;
-        }
-        vec.push_back(buckets_[idx].second);
-        if (duplicate_num == 0) {
-          return vec;
-        }
-      } else if (buckets_[idx].first.is_not_null == false ||
-                 (vec.size() == duplicate_num + unsigned(1))) {
-        return vec;
-      }
-    }
-    return vec;
-  }
+  bool contains_impl(const K& key);
 
   template <typename K>
-  size_t key_to_idx(const K& key) const noexcept(noexcept(hasher()(key))) {
-    const size_t mask = buckets_.size() - 1;
-    return hasher()(key) & mask;
-  }
+  mapped_type find_impl(const K& key);
 
-  size_t probe_next(size_t idx) const noexcept {
-    const size_t mask = buckets_.size() - 1;
-    return (idx + 1) & mask;
-  }
+  template <typename K>
+  std::vector<mapped_type> find_all_impl(const K& key);
+
+  template <typename K>
+  size_t key_to_idx(const K& key) const noexcept(noexcept(hasher()(key)));
+
+  size_t probe_next(size_t idx) const noexcept;
 
  private:
   key_type empty_key_;
   buckets buckets_;
   size_t size_ = 0;
 };
 }  // namespace cider_hashtable
+
+#include <exec/operator/join/CiderLinearProbingHashTable.cpp>
```

Comment:
Better to use #include "exec/operator/join/CiderLinearProbingHashTable.cpp" for custom files